File: rotate-list/rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {


 

public:
    ListNode* rotateRight(ListNode* head, int k) {

        int len =0 ;
       ListNode* temp=head;

     while(temp!=NULL){
         len++;
         temp=temp->next;
     }
       temp=head;
       ListNode* prev=NULL;

      if(head==NULL || head->next==NULL || k%len==0)
      return head;

        k = k % len;
        int i=0;
        int t= len-k;

        while(t--){
          prev=temp;
          temp=temp->next;
        }

        prev->next = NULL;
        ListNode* tail = temp;

        while(tail!=NULL && tail->next!=NULL){
            tail=tail->next;
        }

        tail->next = head;
        
       return temp;
        
    }
};

File: 61-rotate-list/rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
    
        if(head==null)
        return head;


        ListNode newhead,tail;

        newhead = tail = head;

        int length = 1;
        
        while(tail.next!=null)
        {
            length++;
            tail=tail.next;
        }
          
          tail.next = head;

        k=k%length; 

        if(k!=0){
            for(int i=0;i<length-k;i++)
            tail = tail.next;
        }

        head = tail.next;
        tail.next = null;
        return head;

        
    }
}